packages/contracts/src/namespaces/root/systems/PaymasterSystem.sol

```diff
@@ -88,37 +83,21 @@ contract PaymasterSystem is System, IPaymaster {
   ) public override {
     _requireFromEntryPoint();
 
-    (address user, uint256 fromAllowance, uint256 fromBalance) = abi.decode(context, (address, uint256, uint256));
+    (address user, uint256 fromAllowance) = abi.decode(context, (address, uint256));
 
     uint256 totalGasCost = actualGasCost + FIXED_POST_OP_GAS * actualUserOpFeePerGas;
-    uint256 allowanceRefund; // always non-negative
-    int256 balanceDiff; // may be negative if FIXED_POST_OP_GAS made the real cost exceed the deducted balance+allowance
 
     if (totalGasCost > fromAllowance) {
-      // If FIXED_POST_OP_GAS made the total cost exceed the deducted balance+allowance,
-      // we attempt to deduct the missing amount from the balance below.
-      balanceDiff = int256(fromAllowance + fromBalance) - int256(totalGasCost);
-    } else {
-      allowanceRefund = fromAllowance - totalGasCost;
-      balanceDiff = int256(fromBalance);
-    }
-
-    if (balanceDiff != 0) {
+      AllowanceLib.spendAllowance(user, fromAllowance);
+      uint256 fromBalance = totalGasCost - fromAllowance;
       uint256 currentBalance = Balance._get(user);
-      int256 newBalance = int256(currentBalance) + balanceDiff;
+      int256 newBalance = int256(currentBalance) - int256(fromBalance);
       if (newBalance < 0) {
-        revert PaymasterSystem_InsufficientFunds(
-          user,
-          totalGasCost,
-          Allowance._get(user) + fromAllowance,
-          currentBalance + fromBalance
-        );
+        revert PaymasterSystem_InsufficientFunds(user, totalGasCost, fromAllowance, currentBalance);
       }
       Balance._set(user, uint256(newBalance));
-    }
-
-    if (allowanceRefund > 0) {
-      Allowance._set(user, Allowance._get(user) + allowanceRefund);
+    } else {
+      AllowanceLib.spendAllowance(user, totalGasCost);
```

Comment:
sanity check for missingAllowance == 0. Could also be cleaner to revert instead of returning the missing amount